### ISSUE TITLE
Meshlab from snap won't load on Ubuntu 16.04 

### ISSUE THREAD

**Issue by meder411** [cnr-isti-vclab/meshlab#271]
There's [another active thread](https://github.com/cnr-isti-vclab/meshlab/issues/215) on here with a similar problem, but they don't report the exact same circumstances and the workaround they suggest doesn't work for me, hence the new post.

I have tried donwloading both `meshlab` and `meshlab-mardy` from the Ubuntu Software Center and tried `snap install  cYdtAIYBmTS1b7L2q8yrMdpAZ38qLv6Q_4.snap` from the download on the Meshlab website. In all cases, they produce the below output:

```
No schema files found: doing nothing.
Gtk-Message: Failed to load module "overlay-scrollbar"
Gtk-Message: Failed to load module "gail"
Gtk-Message: Failed to load module "atk-bridge"
Gtk-Message: Failed to load module "unity-gtk-module"
Gtk-Message: Failed to load module "canberra-gtk-module"
```

and they also open a small window tiles MeshLab_64bit_fp which hangs for a few seconds before going gray and staying unresponsive. As recently as August, the snap install worked on my machine. Using the old version from apt-get still works in the meantime, but the 2016 update is so much better.

I am running Ubuntu 16.04.3 LTS and with the `4.14.0-041400-generic` kernel.

**Comment by elalish** [comment]
Same exact results on my machine. It used to work great; what happened?

**Comment by Jarek-SRT** [comment]
It seems that at the moment the simplest way to build and run meshlab on Linux would be to slightly tweak the Dockerfile prepared by @hamzamerzic and use [this aproach](https://blog.jessfraz.com/post/docker-containers-on-the-desktop/) to run meshlab GUI in docker.

**Comment by hamzamerzic** [comment]
There is no need to run the GUI in Docker. One can simply run the instructions from the Dockerfile in Linux in order to build meshlab from source. This can be done "by omitting ARG and RUN (replacing RUN with sudo where appropriate) and replacing WORKDIR with 'cd'." I have this setup running locally and it works like a charm. After meshlab is built just `cd src/distrib` and run `./meshlab` from there.

**Comment by Jarek-SRT** [comment]
@hamzamerzic you are correct. I was referring to a situation when one cannot simply build from source in their local environment due to unsatisfied dependencies. I'm running a distro with older version of Qt and this distro's system packages have strict dependency on this particular, older version. It means I cannot simply upgrade Qt to the version needed by MeshLab. That's exactly the problem snap was supposed to solve. As it doesn't work, building and running inside docker seems to be an alternative. Hope it's now more clear what I meant.

**Comment by eric-schleicher** [comment]
also note the following [launchpad repo](https://launchpad.net/~zarquon42/+archive/ubuntu/meshlab)  this has been the best meshlab experience for me.

**Comment by mcallieri** [comment]
Hi all.
the SNAP package was created with the help of one helpful guy from the community.
We are not really expert in Linux distribution. 
We have some colleagues, now, looking into the matter.
However, if
- do you think SNAP package is a good way to distribute MeshLab for Linux?
- which other method should we support?
We have an automatic build tool for Linux, on TRAVIS, but we do not know how to turn it into an automatic installer/package creator.
Any suggestion is welcome.

**Comment by Jarek-SRT** [comment]
Personally I think it would be worth to switch from Snap to AppImage (see https://appimage.org/) It's a similar concept, but has better support across different Linux distributions. Ultimaker Cura is one of many examples of software packages, which have Linux versions distributed in AppImage format.

**Comment by john-davies** [comment]
I've only used one AppImage package ( for LMMS - https://lmms.io/ ) but it works very well.

**Comment by john-davies** [comment]
@hamzamerzic, thank you very much for the build instructions. I followed these and it made the build process very easy. Some brief notes which may help others:

1. Running the apt-get update and upgrade on the same line gave me errors referring to lock files. When I ran them separately it was fine. This may be an artefact of my system.

2. I forked the latest version of the meshlab and vcglib repos rather than using @hamzamerzic's forks. _A reminder that the vcglib needs to be on the DEVEL branch ( See issue #258 - yeah I didn't read the build instructions either :) )_

3. The "QMAKE_FLAGS" line has an absolute path:

    `LIBS+=-L/meshlab/src/external/lib/linux-g++` 
    which needs to be something like:
    `LIBS+=-L/<YOUR PATH HERE>/meshlab/src/external/lib/linux-g++`

4. As per the fix in issue #215 I then replaced the old meshlab executable in /usr/bin with a simple script file that runs this locally built version. This has the advantage of keeping all the file associations.

5. When you run this locally built version for the first time you get an error about version numbers. I assume that this can be safely ignored.

6. I had errors when building `meshlab_full.pro`, I fixed by commenting out the following plugins. _I assume that it's a case of supplying the correct libraries as listed and rebuilding. For now I think that I have sufficient functionality for my needs so this is a low priority task._

    io_3ds - lib3ds.a
    io_ctm - libopenctm.a _( package libopenctm1 is installed )_
    filter_func - libmuparser.a
    filter_isoparametrization - isoparametrization.so
    filter_ssynth - libssynth.a
    edit_mutualcorrs - libedit_mutualcorrs.so

**Comment by John-Nagle** [comment]
On Ubuntu 16.04 LTS, I can run the GUI version of Meshlab from the "snap" of version 2016.12. But trying to run "meshlab.meshlabserver" produces

```
snap run meshlab.meshlabserver -d /tmp/infofile.txt
Gtk-Message: Failed to load module "overlay-scrollbar"
Gtk-Message: Failed to load module "gail"
Gtk-Message: Failed to load module "atk-bridge"
Gtk-Message: Failed to load module "unity-gtk-module"
Gtk-Message: Failed to load module "canberra-gtk-module"
Qt: Session management error: None of the authentication protocols specified are supported
Loading Plugins:
Current Plugins Dir is: /snap/meshlab/4/bin/plugins 
QNetworkManagerInterface::QNetworkManagerInterface(QObject*) propsReply "An AppArmor policy prevents this sender from sending this message to this recipient; type=\"method_call\", sender=\":1.217\" (uid=1000 pid=30231 comm=\"meshlabserver -d /tmp/infofile.txt \") interface=\"org.freedesktop.DBus.Properties\" member=\"GetAll\" error name=\"(unset)\" requested_reply=\"0\" destination=\"org.freedesktop.NetworkManager\" (uid=0 pid=925 comm=\"/usr/sbin/NetworkManager --no-daemon \")"
QNetworkManagerInterface::QNetworkManagerInterface(QObject*) nmReply "An AppArmor policy prevents this sender from sending this message to this recipient; type=\"method_call\", sender=\":1.217\" (uid=1000 pid=30231 comm=\"meshlabserver -d /tmp/infofile.txt \") interface=\"org.freedesktop.NetworkManager\" member=\"GetDevices\" error name=\"(unset)\" requested_reply=\"0\" destination=\"org.freedesktop.NetworkManager\" (uid=0 pid=925 comm=\"/usr/sbin/NetworkManager --no-daemon \")"
"Object path cannot be empty"
"Cannot load library /snap/meshlab/4/bin/plugins/libfilter_ssynth.so: (/snap/meshlab/4/bin/plugins/libfilter_ssynth.so: undefined symbol: _ZTIN14StructureSynth5Model9Rendering16TemplateRendererE)"
Total 239 filtering actions
Total 11 io plugins
called here!
Segmentation fault (core dumped)
```

I was writing a Blender plug-in which uses Meshlab Server. But I can't do that when Meshlab crashes and every user reports a different crash. The whole point of "snaps" is that they are supposed to Just Work.

I can run the older version of Meshlab Server from the Ubuntu repository, but it's an older version and the quadric mesh reduction filter crashes. If you can't fix the "snap", please put a current version in the regular Ubuntu repository, as you did for previous versions. Thanks.

**Comment by eric-schleicher** [comment]
just a (reminder) pointer again to the ppa that has a build of the latest meshlab release.  if you're here struggling with the snap on ubuntu, consider using the PPA listed in the above comment.

https://github.com/cnr-isti-vclab/meshlab/issues/271#issuecomment-369700074

**Comment by John-Nagle** [comment]
Tried the version from

**Comment by John-Nagle** [comment]
The problem is in **libfilter_layer.so.**  The code that lists the filters does this by linking to each .so file and running a function which returns the filter's description in XML. When the filter lister gets to libfilter_layer.so, that .so's "InitParameterSet" function crashes. If I move libfilter_layer.so out of the plugins directory, the crash does not happen.

**Comment by John-Nagle** [comment]
I also get the message "Error While parsing the XML filter plugin descriptors: We are trying to load a xml file that does not correspond to any dll or javascript code; please delete all the spurious xml files." Suggest finding what prints that message and putting the file name of the problem file in the message.

**Comment by John-Nagle** [comment]
Looking at the source code in "https://github.com/cnr-isti-vclab/meshlab/blob/master/src/meshlabplugins/filter_layer/filter_layer.cpp", we can see the problem. At 

```
void FilterLayerPlugin::initParameterSet(QAction *action, MeshDocument &md, RichParameterSet & parlst)
{
    MeshModel *mm=md.mm();
    RasterModel *rm=md.rm();
```
...
```
    case FP_RENAME_MESH :
        parlst.addParam(new RichString ("newName",
            mm->label(),
            "New Label",
            "New Label for the mesh"));
        break;
    case FP_RENAME_RASTER :
        parlst.addParam(new RichString ("newName",
            rm->label(),
            "New Label",
            "New Label for the raster"));
        break;
```

When the plug-ins are just being listed, there's no mesh model or raster model. So dereferencing "mm" or "rm" results in a dereference of null and a segfault. This doesn't happen in the GUI program because the menu which would display those features is greyed out until you load a mesh. 

I'm not familiar enough with this code to submit a patch, but clearly, if "mm" or "rm" is null, don't  dereference it.. Yet another classic C/C++ "but I didn't know it could be null" bug. 

Happy Thanksgiving.

**Comment by constructive-realities** [comment]
this is all sort of leading the the conclusion that the SNAPs are junk and that a random PPA that isn't maintained for new ubuntu releases isn't a reliable way to distribute MeshLab.  Given how critical it is to so many people (myself includes) i would love to see it setup on a service like https://www.appveyor.com/ where everyone would have access to builds from a continuous integration systems.   

It's free for open source projects.

**Comment by John-Nagle** [comment]
The problem I found above has nothing to do with the packaging process. It's a ordinary C bug.

The input data parsing is very brittle. Make a mistake in an XML script for "meshlabserver", and it usually segfaults.

I'm writing an add-on for Blender to do mesh reduction using "meshlabserver", and for that I need a reliable distribution of meshlab on Linux, Windows, and Mac.

**Comment by martin-frbg** [comment]
@John-Nagle  looks to me as if `if (mm)` (and 'if (rm)' respectively) before the parlst.addParam should be sufficient.
@constructive-realities one of the project owners would need to set up the connection to the build service (unless someone wants to try it from their own fork of meshlab). As they seem to be already using travis to autobuild  dmg images for osx with each point release, I suspect the primary objective would be to debug the current snap build, see how to integrate it with travis (I just found a tutorial at https://tutorials.ubuntu.com/tutorial/continuous-snap-delivery-from-travis-ci) and submit a PR with the augmented .travis.yml file (if it fits the time limits of free travis CI builds).

**Comment by alemuntoni** [comment]
This issue should be solved in the latest version (2020.06). Feel free to reopen if the problem persists.
